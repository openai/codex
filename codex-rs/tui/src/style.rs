use crate::color::blend;
use crate::color::is_light;
use crate::color::perceptual_distance;
use crate::terminal_palette::basic_palette;
use crate::terminal_palette::terminal_palette;
use ratatui::style::Color;
use ratatui::style::Style;

/// Returns the style for a user-authored message using the provided terminal background.
pub fn user_message_style(terminal_bg: Option<(u8, u8, u8)>) -> Style {
    match terminal_bg {
        Some(bg) => Style::default().bg(user_message_bg(bg)),
        None => Style::default(),
    }
}

#[allow(clippy::disallowed_methods)]
pub fn user_message_bg(terminal_bg: (u8, u8, u8)) -> Color {
    // Determine a reference "top" color to blend toward. For dark backgrounds we lighten
    // the mix with white, and for light backgrounds we darken it with black.
    let top = if is_light(terminal_bg) {
        (0, 0, 0)
    } else {
        (255, 255, 255)
    };
    let bottom = terminal_bg;
    let Some(mut color_level) = supports_color::on_cached(supports_color::Stream::Stdout) else {
        return Color::default();
    };

    #[cfg(windows)]
    // Windows Terminal has been the default shell application for Windows since October 2022
    // and has supported truecolor even longer. However it usually does not set COLORTERM to indicate that.
    // so this is a pretty safe heuristic.
    if std::env::var_os("WT_SESSION").is_some() {
        color_level.has_16m = true;
    }

    let target = blend(top, bottom, 0.1);
    if color_level.has_16m {
        // In truecolor terminals we can use the exact RGB value.
        let (r, g, b) = target;
        Color::Rgb(r, g, b)
    } else if color_level.has_256
        && let Some(palette) = terminal_palette()
        && let Some((i, _)) = palette.into_iter().enumerate().min_by(|(_, a), (_, b)| {
            perceptual_distance(*a, target)
                .partial_cmp(&perceptual_distance(*b, target))
                .unwrap_or(std::cmp::Ordering::Equal)
        })
    {
        // If we have a captured 256-color palette, pick whichever indexed color is
        // perceptually closest to the blended target.
        Color::Indexed(i as u8)
    } else if color_level.has_basic {
        if let Some(palette) = basic_palette() {
            // On Windows terminals the palette is configurable, so evaluate the actual
            // runtime color table to keep the blended shading aligned with custom themes.
            closest_runtime_basic_color(target, terminal_bg, &palette)
        } else {
            // Finally, degrade to the well-known ANSI 16-color defaults using a perceptual
            // distance match.
            closest_basic_color(target, terminal_bg)
        }
    } else {
        Color::default()
    }
}

fn closest_runtime_basic_color(
    target: (u8, u8, u8),
    terminal_bg: (u8, u8, u8),
    palette: &[(u8, u8, u8); 16],
) -> Color {
    select_basic_palette_color(
        target,
        terminal_bg,
        palette.iter().enumerate().filter_map(|(idx, rgb)| {
            BASIC_TERMINAL_COLORS
                .get(idx)
                .map(|(color, _)| (*rgb, *color))
        }),
    )
}

fn closest_basic_color(target: (u8, u8, u8), terminal_bg: (u8, u8, u8)) -> Color {
    // Iterate through the baked-in ANSI colors and return whichever one is closest to the
    // desired RGB shade while maintaining contrast with the background.
    select_basic_palette_color(
        target,
        terminal_bg,
        BASIC_TERMINAL_COLORS
            .iter()
            .map(|(color, rgb)| (*rgb, *color)),
    )
}

const MIN_PERCEPTUAL_DISTANCE: f32 = 6.0;

fn select_basic_palette_color(
    target: (u8, u8, u8),
    terminal_bg: (u8, u8, u8),
    entries: impl Iterator<Item = ((u8, u8, u8), Color)>,
) -> Color {
    let mut best = None;
    let mut fallback = None;
    for (rgb, color) in entries {
        let dist = perceptual_distance(rgb, target);
        if fallback.is_none_or(|(_, best_dist)| dist < best_dist) {
            fallback = Some((color, dist));
        }
        if perceptual_distance(rgb, terminal_bg) > MIN_PERCEPTUAL_DISTANCE
            && best.is_none_or(|(_, best_dist)| dist < best_dist)
        {
            best = Some((color, dist));
        }
    }
    best.or(fallback)
        .map(|(color, _)| color)
        .unwrap_or(Color::default())
}

// Mapping of ANSI color indices to approximate RGB tuples. These values mirror Windows'
// default console palette so the fallback path stays visually consistent across platforms.
const BASIC_TERMINAL_COLORS: [(Color, (u8, u8, u8)); 16] = [
    (Color::Black, (0, 0, 0)),
    (Color::Blue, (0, 0, 128)),
    (Color::Green, (0, 128, 0)),
    (Color::Cyan, (0, 128, 128)),
    (Color::Red, (128, 0, 0)),
    (Color::Magenta, (128, 0, 128)),
    (Color::Yellow, (128, 128, 0)),
    (Color::Gray, (192, 192, 192)),
    (Color::DarkGray, (128, 128, 128)),
    (Color::LightBlue, (0, 0, 255)),
    (Color::LightGreen, (0, 255, 0)),
    (Color::LightCyan, (0, 255, 255)),
    (Color::LightRed, (255, 0, 0)),
    (Color::LightMagenta, (255, 0, 255)),
    (Color::LightYellow, (255, 255, 0)),
    (Color::White, (255, 255, 255)),
];
