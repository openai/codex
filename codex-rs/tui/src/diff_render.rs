use ratatui::style::Color;
use ratatui::style::Modifier;
use ratatui::style::Style;
use ratatui::text::Line as RtLine;
use ratatui::text::Span as RtSpan;
use std::collections::HashMap;
use std::path::PathBuf;

use codex_core::protocol::FileChange;

struct FileSummary {
    display_path: String,
    added: usize,
    removed: usize,
}

pub(crate) fn create_diff_summary(
    title: &str,
    changes: HashMap<PathBuf, FileChange>,
) -> Vec<RtLine<'static>> {
    let mut files: Vec<FileSummary> = Vec::new();

    // Count additions/deletions from a unified diff body
    let count_from_unified = |diff: &str| -> (usize, usize) {
        if let Ok(patch) = diffy::Patch::from_str(diff) {
            let mut adds = 0usize;
            let mut dels = 0usize;
            for hunk in patch.hunks() {
                for line in hunk.lines() {
                    match line {
                        diffy::Line::Insert(_) => adds += 1,
                        diffy::Line::Delete(_) => dels += 1,
                        _ => {}
                    }
                }
            }
            (adds, dels)
        } else {
            let mut adds = 0usize;
            let mut dels = 0usize;
            for l in diff.lines() {
                if l.starts_with("+++") || l.starts_with("---") || l.starts_with("@@") {
                    continue;
                }
                match l.as_bytes().first() {
                    Some(b'+') => adds += 1,
                    Some(b'-') => dels += 1,
                    _ => {}
                }
            }
            (adds, dels)
        }
    };

    for (path, change) in &changes {
        use codex_core::protocol::FileChange::*;
        match change {
            Add { content } => {
                let added = content.lines().count();
                files.push(FileSummary {
                    display_path: path.display().to_string(),
                    added,
                    removed: 0,
                });
            }
            Delete => {
                let removed = std::fs::read_to_string(path)
                    .ok()
                    .map(|s| s.lines().count())
                    .unwrap_or(0);
                files.push(FileSummary {
                    display_path: path.display().to_string(),
                    added: 0,
                    removed,
                });
            }
            Update {
                unified_diff,
                move_path,
            } => {
                let (added, removed) = count_from_unified(unified_diff);
                let display_path = if let Some(new_path) = move_path {
                    format!("{} → {}", path.display(), new_path.display())
                } else {
                    path.display().to_string()
                };
                files.push(FileSummary {
                    display_path,
                    added,
                    removed,
                });
            }
        }
    }

    let file_count = files.len();
    let total_added: usize = files.iter().map(|f| f.added).sum();
    let total_removed: usize = files.iter().map(|f| f.removed).sum();
    let noun = if file_count == 1 { "file" } else { "files" };

    let mut out: Vec<RtLine<'static>> = Vec::new();

    // Header
    let mut header_spans: Vec<RtSpan<'static>> = Vec::new();
    header_spans.push(RtSpan::styled(
        title.to_owned(),
        Style::default()
            .fg(Color::Magenta)
            .add_modifier(Modifier::BOLD),
    ));
    header_spans.push(RtSpan::raw(" to "));
    header_spans.push(RtSpan::raw(format!("{file_count} {noun} ")));
    header_spans.push(RtSpan::raw("("));
    header_spans.push(RtSpan::styled(
        format!("+{total_added}"),
        Style::default().fg(Color::Green),
    ));
    header_spans.push(RtSpan::raw(" "));
    header_spans.push(RtSpan::styled(
        format!("-{total_removed}"),
        Style::default().fg(Color::Red),
    ));
    header_spans.push(RtSpan::raw(")"));
    out.push(RtLine::from(header_spans));

    // Dimmed per-file lines with prefix
    for (idx, f) in files.iter().enumerate() {
        let mut spans: Vec<RtSpan<'static>> = Vec::new();
        spans.push(RtSpan::raw(f.display_path.clone()));
        spans.push(RtSpan::raw(" ("));
        spans.push(RtSpan::styled(
            format!("+{}", f.added),
            Style::default().fg(Color::Green),
        ));
        spans.push(RtSpan::raw(" "));
        spans.push(RtSpan::styled(
            format!("-{}", f.removed),
            Style::default().fg(Color::Red),
        ));
        spans.push(RtSpan::raw(")"));

        let mut line = RtLine::from(spans);
        let prefix = if idx == 0 { "  ⎿ " } else { "    " };
        line.spans.insert(0, prefix.into());
        line.spans.iter_mut().for_each(|span| {
            span.style = span.style.add_modifier(Modifier::DIM);
        });
        out.push(line);
    }

    out
}
