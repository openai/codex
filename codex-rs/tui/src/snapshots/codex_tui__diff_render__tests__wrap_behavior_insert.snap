---
source: tui/src/diff_render.rs
assertion_line: 380
expression: terminal.backend()
---
"    1     +this is a very long line that should wrap across multiple terminal col         "
"           umns and continue                                                              "
"                                                                                          "
"                                                                                          "
"                                                                                          "
"                                                                                          "
"                                                                                          "
"                                                                                          "
